syntax = "proto3";
package kvrf;

message PutRequest{
    string key = 1;
    string value = 2;
    string op = 3;
    int32 clientId = 4;
    int32 requestId = 5;
}

message PutResponse{
    bool isWrongGroup = 1;
    bool isWrongLeader = 2;
    bool OK = 3;
}

message GetRequest{
    string key = 1;
    int32 clientId = 2;
    int32 requestId = 3;
}

message GetResponse{
    bool isWrongGroup = 1;
    bool isWrongLeader = 2;
    string value = 3;
    bool isExist = 4;
}

message Request{
    bytes type = 1;
    bytes request_msg = 2;
}

message Response{
    bytes type = 1;
    bool isWrongLeader = 2;
    bytes response_msg = 3;
}

message MigrationRequest{
    int32 shardid = 1;
    int32 configid = 2;
}

message MigrationResponse{
    bytes err = 1;
    int32 shardid = 2;
    int32 configid = 3;
    map<string , bytes> database = 4;
    map<int32 , int32> clientreq = 5;
}

message AckRequest{
    int32 configid = 1;
    int32 shardid = 2;
}

message AckResponse{
    bool received = 1;
    bytes error = 2;
}
//kvshard

service KvRaftServiceRpc{
    rpc Put(PutRequest) returns(PutResponse);
    rpc Get(GetRequest) returns(GetResponse);
    rpc Migrate(MigrationRequest) returns(MigrationResponse);
    rpc Ack(AckRequest) returns(AckResponse);
}